### ISSUE TITLE
How should for-each behave if one of the commands fails?

### ISSUE THREAD

**Issue by p-jackson** [JohnRGee/forest-arborist#2]
I wanted to push all the repos in my forest (well, at least all the repos I have write access to).

I tried `fab for-each git push`, but when I got to the repo I don't have write access to it stopped and didn't run `git push` on the rest of the repos.

Similar thing would happen if one of my repos was a mercurial repo I think.

What if `for-each` didn't stop after an error, but saved up the error code so if any of the commands returned non-zero then fab would return non-zero?

**Comment by shadowspawn** [comment]
I had realised that "hg push" with nothing to push returns 1 status making `fab for-each hg push` problematic, but decided to wait for more real world examples. Like yours. :-)

I am not sure that continuing by default is a good idea, since if you did a tragic typo it would be nicer to stop when it explodes the first time than repeat it across forest.

A possible pattern is to follow the make utility which has support for ignoring errors, and has been around for a long time:
```make --keep-going```

**Comment by shadowspawn** [comment]
(And as for git/hg support, I was assuming either/or rather than mixed. However, I had mused about options on for-each like git+hg and free+locked+pinned, for tuning the processed repos in complex forests. Again, waiting to see if useful in practice rather than in theory.)

**Comment by p-jackson** [comment]
Yes, I think opt in is good 👍 
Using git _and_ hg isn't something we really need, we're probably just doing it because we got away with it. It's not really a killer feature IMO so don't think it should really be a focus.

**Comment by shadowspawn** [comment]
Added "--keepgoing" for for-each and for-free.

[0cd9fe850532e7218fe76cd795cb4c2ed2d117fd](https://github.com/JohnRGee/forest-arborist/commit/0cd9fe850532e7218fe76cd795cb4c2ed2d117fd)

**Comment by p-jackson** [comment]
Tested on Mac and Windows, both the "one of my repos is hg" and "one of my repos is readonly" cases.

I really like the `(to keep going despite errors use "fab for-each --keepgoing")` message. First time I saw it though was because I mistyped the command, so `--keepgoing` wasn't the right hint.

Another interesting thing I discovered: on Win when I didn't have permission to `git push` a repo the command returns non-zero; on Mac it prompts you for a username and password. So the feature doesn't _quite_ fix my use case on Mac.

I still like it though, I think I'll continue to use the feature.
